cpp/scenario/gazebo/include/scenario/gazebo/Joint.h

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2020 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This project is dual licensed under LGPL v2.1+ or Apache License.
+ *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SCENARIO_GAZEBO_JOINT_H
+#define SCENARIO_GAZEBO_JOINT_H
+
+#include <ignition/gazebo/Entity.hh>
+#include <ignition/gazebo/EntityComponentManager.hh>
+#include <ignition/gazebo/EventManager.hh>
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace scenario {
+    namespace base {
+        struct PID;
+        struct Limit;
+        enum class JointType
+        {
+            Invalid,
+            Fixed,
+            Revolute,
+            Prismatic,
+            Ball,
+        };
+        enum class JointControlMode
+        {
+            Idle,
+            Force,
+            Velocity,
+            Position,
+            PositionInterpolated,
+        };
+    } // namespace base
+    namespace gazebo {
+        class Joint;
+        class Model;
+    } // namespace gazebo
+} // namespace scenario
+
+class scenario::gazebo::Joint
+{
+public:
+    Joint();
+    virtual ~Joint();
+
+    // ============
+    // Gazebo Joint
+    // ============
+
+    uint64_t id() const;
+    bool initialize(const ignition::gazebo::Entity jointEntity,
+                    ignition::gazebo::EntityComponentManager* ecm,
+                    ignition::gazebo::EventManager* eventManager);
+    bool createECMResources();
+
+    bool historyOfAppliedJointForcesEnabled() const;
+    bool enableHistoryOfAppliedJointForces( //
+        const bool enable,
+        const size_t maxHistorySize = 100);
+    std::vector<double> getHistoryOfAppliedJointForces() const;
```

Comment:
Just out of curiosity, could you give an example use case needing the history of applied joint forces?

Reply:
We already encountered  in the past a use case where these methods are fundamental. Consider the following case:

- You setup Gazebo to perform multiple physics step every time you call `GazeboSimulator::run`.
- You have a joint controller (e.g. when you're controlling joints in `JointControlMode::Position` there is one PID for each joint).
- You want to extract the generalized joint forces that your controllers applied.

From your user code, after you step the simulator you can extract only the very last torque that the PID controllers applied to the model. Instead, these methods allow to collect all of them.

In our case, the setup could be comparable of an environment that is stepped by an agent with a lower rate than the physics (e.g. 1 environment step = 50 physics steps). In many cases, you want to minimize the control effort, and if you cannot get a window of values you're going to lose the first 49 values.